File: packages/nodx-js/src/nods.mjs
import { DEFAULT_LIMITS } from "./limits.mjs";
import { classifyUri, ReferenceKind } from "./url.mjs";

// JavaScript port of nodx-style. Mirrors the Rust auditor for parity.

const ALLOWED_PROPERTIES = new Set([
  "background", "background-color", "background-image", "background-position",
  "background-repeat", "background-size", "border", "border-block",
  "border-block-end", "border-block-start", "border-bottom", "border-collapse",
  "border-color", "border-inline", "border-inline-end", "border-inline-start",
  "border-left", "border-radius", "border-right", "border-spacing", "border-style",
  "border-top", "border-width", "box-decoration-break", "box-shadow",
  "break-after", "break-before", "break-inside", "color", "column-count",
  "column-gap", "column-width", "columns", "direction", "display", "flex",
  "flex-basis", "flex-direction", "flex-grow", "flex-shrink", "flex-wrap",
  "font", "font-family", "font-feature-settings", "font-size", "font-style",
  "font-variant", "font-weight", "gap", "grid-column", "grid-column-gap",
  "grid-row", "grid-row-gap", "grid-template-columns", "grid-template-rows",
  "hanging-punctuation", "height", "hyphens", "justify-content", "justify-items",
  "letter-spacing", "line-height", "list-style", "list-style-position",
  "list-style-type", "margin", "margin-block", "margin-block-end",
  "margin-block-start", "margin-bottom", "margin-inline", "margin-inline-end",
  "margin-inline-start", "margin-left", "margin-right", "margin-top",
  "max-height", "max-width", "min-height", "min-width", "opacity", "orphans",
  "outline", "outline-color", "outline-offset", "outline-style", "outline-width",
  "overflow", "overflow-wrap", "padding", "padding-block", "padding-block-end",
  "padding-block-start", "padding-bottom", "padding-inline", "padding-inline-end",
  "padding-inline-start", "padding-left", "padding-right", "padding-top",
  "page-break-after", "page-break-before", "page-break-inside", "position",
  "quotes", "size", "tab-size", "table-layout", "text-align", "text-decoration",
  "text-decoration-color", "text-decoration-style", "text-decoration-thickness",
  "text-indent", "text-transform", "text-underline-offset", "vertical-align",
  "white-space", "widows", "width", "word-break", "word-spacing", "writing-mode",
]);

const ALLOWED_DISPLAY = new Set([
  "block", "inline", "inline-block", "list-item", "table", "table-row",
  "table-cell", "table-header-group", "table-row-group", "table-footer-group",
  "none", "flex", "inline-flex", "grid", "inline-grid",
]);

const ALLOWED_POSITION = new Set(["static", "relative"]);

const SAFE_TAGS = new Set([
  "a", "article", "aside", "blockquote", "body", "caption", "code", "dd", "div",
  "dl", "dt", "em", "figcaption", "figure", "h1", "h2", "h3", "h4", "h5", "h6",
  "hr", "html", "img", "li", "main", "mark", "nav", "ol", "p", "pre", "section",
  "span", "strong", "sub", "sup", "table", "tbody", "td", "th", "thead", "tr",
  "ul", "var",
]);

const STRUCTURAL_PSEUDOS = new Set([
  ":root", ":first-child", ":last-child", ":only-child", ":empty",
]);

export function auditStylesheet(input, limits = DEFAULT_LIMITS) {
  const violations = [];
  auditBreakouts(input, violations);
  for (const rule of parseRules(input)) {
    auditRule(rule, limits, violations);
  }
  return dedupe(violations);
}

export function sanitizeStylesheet(input, limits = DEFAULT_LIMITS) {
  const audit = auditStylesheet(input, limits);
  if (audit.length === 0) return escapeStyleText(input);
  if (audit.some((violation) => violation.message === "Forbidden executable or breakout content in NODS.")) {
    return "/* NODX-E027: blocked unsafe style content */";
  }
  const out = [];
  for (const rule of parseRules(input)) {
    const ruleAudit = [];
    auditRule(rule, limits, ruleAudit);
    if (ruleAudit.some((violation) => violation.severity === "error")) {
      out.push("/* NODX-E027: forbidden NODS rule omitted */");
    } else {
      out.push(escapeStyleText(rule.source));
    }
  }
  return out.length ? out.join("") : "/* NODX-E027: blocked unsafe style content */";
}

export function yamlStyleToCss(input) {
  const rules = [];
  let selector = "";
  let declarations = [];
  for (const raw of input.split(/\r?\n/)) {
    if (!raw.trim() || raw.trimStart().startsWith("#")) continue;
    if (/^\t/.test(raw)) throw new Error("YAML style indentation must use spaces.");
    const indent = raw.length - raw.trimStart().length;
    const line = raw.trimEnd();
    if (indent === 0) {
      flushRule();
      if (!line.endsWith(":")) throw new Error("YAML style top-level entries must end with `:`.");
      selector = line.slice(0, -1).trim();
      if (!selector) throw new Error("Invalid YAML style selector.");
    } else if (selector) {
      const i = line.indexOf(":");
      if (i <= 0) throw new Error("YAML style declaration must use `property: value`.");
      const property = line.slice(0, i).trim();
      const value = line.slice(i + 1).trim();
      if (!property || !value) throw new Error("YAML style declarations require a property and value.");
      declarations.push(`${property}: ${value};`);
    } else {
      throw new Error("Invalid YAML style structure.");
    }
  }
  flushRule();
  return rules.join("\n");

  function flushRule() {
    if (selector && declarations.length) rules.push(`${selector} { ${declarations.join(" ")} }`);
    selector = "";
    declarations = [];
  }
}

function parseRules(input) {
  const rules = [];
  let start = 0;
  while (start < input.length) {
    const open = input.indexOf("{", start);
    if (open < 0) break;
    let depth = 1;
    let i = open + 1;
    while (i < input.length && depth > 0) {
      const ch = input[i];
      if (ch === "{") depth += 1;
      else if (ch === "}") depth -= 1;
      i += 1;
    }
    if (depth !== 0) {
      rules.push({
        prelude: input.slice(start, open),
        declarations: input.slice(open + 1),
        source: input.slice(start),
      });
      break;
    }
    rules.push({
      prelude: input.slice(start, open),
      declarations: input.slice(open + 1, i - 1),
      source: input.slice(start, i),
    });
    start = i;
  }
  return rules;
}

function auditBreakouts(input, violations) {
  const decoded = decodeCssEscapes(input).toLowerCase();
  for (const c of [
    "</style", "<script", "<svg", "<iframe", "<object", "<embed", "vbscript:",
    "expression(", "@import",
  ]) {
    if (decoded.includes(c)) {
      violations.push(err(c, "Forbidden executable or breakout content in NODS."));
    }
  }
}

function auditRule(rule, limits, violations) {
  const prelude = rule.prelude.trim();
  if (prelude.startsWith("@")) {
    auditAtRule(prelude, violations);
    const lower = prelude.toLowerCase();
    if (lower.startsWith("@media") || lower.startsWith("@supports")) {
      for (const nested of parseRules(rule.declarations)) {
        auditRule(nested, limits, violations);
      }
      return;
    }
  } else {
    auditSelector(prelude, violations);
  }
  auditDeclarations(rule.declarations, limits, violations);
}

function auditAtRule(prelude, violations) {
  const lower = prelude.toLowerCase();
  if (lower.startsWith("@page") || lower.startsWith("@media") || lower.startsWith("@supports")) return;
  violations.push(err(prelude, "Forbidden NODS at-rule."));
}

function auditSelector(selector, violations) {
  if (selector === "") {
    violations.push(err("selector", "Missing NODS selector."));
    return;
  }
  for (const part of selector.split(",")) {
    const trimmed = part.trim();
    if (trimmed === "") {
      violations.push(err("selector", "Empty NODS selector list entry."));
      continue;
    }
    for (const token of trimmed.split(/\s+/)) {
      auditSelectorToken(token, violations);
    }
  }
}

function auditSelectorToken(token, violations) {
  if (token === "") {
    violations.push(err("selector", "Empty NODS selector token."));
    return;
  }
  if (token === ">" || token === "+" || token === "~") return;
  if (STRUCTURAL_PSEUDOS.has(token)) return;
  if (token.startsWith(":not(") && token.endsWith(")")) {
    auditSelectorToken(token.slice(5, -1), violations);
    return;
  }
  if (token.includes("::")) {
    violations.push(err(token, "Pseudo-elements are not allowed in NODS."));
    return;
  }
  if (token.includes(":")) {
    violations.push(err(token, "Interactive NODS pseudo-class is forbidden."));
    return;
  }
  if (token.includes("*") || token.includes("|")) {
    violations.push(err(token, "Forbidden NODS selector."));
    return;
  }
  const bracketStart = token.indexOf("[");
  if (bracketStart >= 0) {
    const bracketEnd = token.indexOf("]");
    if (bracketEnd < 0) {
      violations.push(err(token, "Unterminated attribute selector."));
      return;
    }
    const base = token.slice(0, bracketStart);
    const attr = token.slice(bracketStart + 1, bracketEnd);
    if (base !== "" && !isSafeSelectorBase(base)) {
      violations.push(err(base, "Unsupported NODS selector."));
      return;
    }
    if (!isSafeAttributeSelector(attr)) {
      violations.push(err(attr, "Forbidden NODS attribute selector."));
    }
    return;
  }
  if (!isSafeSelectorBase(token)) {
    violations.push(err(token, "Unsupported NODS selector."));
  }
}

function isSafeSelectorBase(token) {
  if (token === "") return false;
  const idx = firstIndexOfAny(token, [".", "#"]);
  if (idx === 0) {
    const kind = token[0];
    const rest = token.slice(1);
    const next = firstIndexOfAny(rest, [".", "#"]);
    const ident = next < 0 ? rest : rest.slice(0, next);
    if (!isIdent(ident)) return false;
    if (next < 0) return kind === "." || kind === "#";
    return isSafeSelectorBase(rest.slice(next));
  }
  if (idx > 0) {
    return isSafeTag(token.slice(0, idx)) && isSafeSelectorBase(token.slice(idx));
  }
  return isSafeTag(token);
}

function isSafeTag(token) {
  return SAFE_TAGS.has(token);
}

function firstIndexOfAny(token, chars) {
  let best = -1;
  for (const ch of chars) {
    const i = token.indexOf(ch);
    if (i >= 0 && (best < 0 || i < best)) best = i;
  }
  return best;
}

function isSafeAttributeSelector(attr) {
  let i = 0;
  let name = "";
  while (i < attr.length) {
    const c = attr[i];
    if (/[A-Za-z0-9_-]/.test(c)) {
      name += c;
      i += 1;
    } else {
      break;
    }
  }
  if (name === "" || !isIdent(name)) return false;
  const rest = attr.slice(i).trim();
  if (rest === "") {
    return ["lang", "dir", "role", "data-tag", "data-tone", "data-variant", "data-color", "data-status"].includes(name);
  }
  for (const op of ["~=", "|=", "^=", "$=", "*=", "="]) {
    if (rest.startsWith(op)) {
      const value = rest.slice(op.length).trim();
      if (
        (value.startsWith('"') && value.endsWith('"') && value.length >= 2) ||
        (value.startsWith("'") && value.endsWith("'") && value.length >= 2)
      ) {
        const inner = value.slice(1, -1);
        return inner.length > 0 && !/["'<>]/.test(inner);
      }
      return false;
    }
  }
  return false;
}

function auditDeclarations(declarations, limits, violations) {
  for (const declaration of declarations.split(";")) {
    const trimmed = declaration.trim();
    if (trimmed === "") continue;
    const idx = trimmed.indexOf(":");
    if (idx < 0) {
      violations.push(err(trimmed, "Malformed NODS declaration."));
      continue;
    }
    const property = trimmed.slice(0, idx).trim().toLowerCase();
    const value = trimmed.slice(idx + 1).trim();
    if (isForbiddenProperty(property)) {
      violations.push(err(property, "Forbidden NODS property."));
      continue;
    }
    if (!isAllowedProperty(property)) {
      violations.push(warn(property, "Unsupported NODS property."));
      continue;
    }
    auditValue(property, value, limits, violations);
  }
}

function isForbiddenProperty(property) {
  return ["behavior", "-moz-binding", "-ms-behavior", "binding"].includes(property);
}

function isAllowedProperty(property) {
  if (property.startsWith("--") && property.length > 2) return true;
  return ALLOWED_PROPERTIES.has(property);
}

function auditValue(property, value, limits, violations) {
  const decoded = decodeCssEscapes(value);
  const lower = decoded.toLowerCase();
  for (const f of ["expression(", "javascript:", "vbscript:", "@import", "behavior:"]) {
    if (lower.includes(f)) {
      violations.push(err(f, "Forbidden NODS construct in value."));
    }
  }
  for (const url of styleUrls(value)) {
    if (!classifyUri(ReferenceKind.Style, url, limits).ok) {
      violations.push(err(url, "Unsafe NODS URL."));
    }
  }
  if (property === "position" && !ALLOWED_POSITION.has(lower)) {
    violations.push(err(value, "Forbidden NODS positioning value."));
  }
  if (property === "display" && !ALLOWED_DISPLAY.has(lower)) {
    violations.push(err(value, "Forbidden NODS display value."));
  }
}

function styleUrls(input) {
  const urls = [];
  const lower = input.toLowerCase();
  let offset = 0;
  while (true) {
    const start = lower.indexOf("url(", offset);
    if (start < 0) break;
    const urlStart = start + 4;
    const end = input.indexOf(")", urlStart);
    if (end < 0) break;
    let inner = input.slice(urlStart, end).trim();
    if (
      (inner.startsWith('"') && inner.endsWith('"')) ||
      (inner.startsWith("'") && inner.endsWith("'"))
    ) {
      inner = inner.slice(1, -1);
    }
    urls.push(inner);
    offset = end + 1;
  }
  return urls;
}

function decodeCssEscapes(input) {
  let out = "";
  let i = 0;
  while (i < input.length) {
    if (input[i] === "\\" && i + 1 < input.length) {
      let j = i + 1;
      let hex = "";
      while (j < input.length && hex.length < 6 && /[0-9a-fA-F]/.test(input[j])) {
        hex += input[j];
        j += 1;
      }
      if (hex.length > 0) {
        if (j < input.length && /[ \t\n]/.test(input[j])) j += 1;
        const code = Number.parseInt(hex, 16);
        if (Number.isFinite(code)) out += String.fromCharCode(code);
        i = j;
        continue;
      }
      if (j < input.length) {
        out += input[j];
        i = j + 1;
        continue;
      }
    }
    out += input[i];
    i += 1;
  }
  return out;
}

function escapeStyleText(input) {
  return input.replace(/</g, "\\3C ").replace(/>/g, "\\3E ");
}

function dedupe(list) {
  const seen = new Set();
  const result = [];
  for (const v of list) {
    const key = v.construct + " " + v.message + " " + v.severity;
    if (!seen.has(key)) {
      seen.add(key);
      result.push(v);
    }
  }
  return result;
}

function err(construct, message) {
  return { construct: construct.trim(), message, severity: "error" };
}

function warn(construct, message) {
  return { construct: construct.trim(), message, severity: "warning" };
}

function isIdent(name) {
  if (name === "") return false;
  const first = name[0];
  if (!(/[A-Za-z_]/.test(first))) return false;
  return /^[A-Za-z_][A-Za-z0-9_-]*$/.test(name);
}

File: packages/nodx-py/src/nodx/nods.py
import re

from .limits import DEFAULT_LIMITS
from .url import ReferenceKind, classify_uri

ALLOWED_PROPERTIES = {
    "background", "background-color", "background-image", "background-position",
    "background-repeat", "background-size", "border", "border-block",
    "border-block-end", "border-block-start", "border-bottom", "border-collapse",
    "border-color", "border-inline", "border-inline-end", "border-inline-start",
    "border-left", "border-radius", "border-right", "border-spacing", "border-style",
    "border-top", "border-width", "box-decoration-break", "box-shadow",
    "break-after", "break-before", "break-inside", "color", "column-count",
    "column-gap", "column-width", "columns", "direction", "display", "flex",
    "flex-basis", "flex-direction", "flex-grow", "flex-shrink", "flex-wrap",
    "font", "font-family", "font-feature-settings", "font-size", "font-style",
    "font-variant", "font-weight", "gap", "grid-column", "grid-column-gap",
    "grid-row", "grid-row-gap", "grid-template-columns", "grid-template-rows",
    "hanging-punctuation", "height", "hyphens", "justify-content", "justify-items",
    "letter-spacing", "line-height", "list-style", "list-style-position",
    "list-style-type", "margin", "margin-block", "margin-block-end",
    "margin-block-start", "margin-bottom", "margin-inline", "margin-inline-end",
    "margin-inline-start", "margin-left", "margin-right", "margin-top",
    "max-height", "max-width", "min-height", "min-width", "opacity", "orphans",
    "outline", "outline-color", "outline-offset", "outline-style", "outline-width",
    "overflow", "overflow-wrap", "padding", "padding-block", "padding-block-end",
    "padding-block-start", "padding-bottom", "padding-inline", "padding-inline-end",
    "padding-inline-start", "padding-left", "padding-right", "padding-top",
    "page-break-after", "page-break-before", "page-break-inside", "position",
    "quotes", "size", "tab-size", "table-layout", "text-align", "text-decoration",
    "text-decoration-color", "text-decoration-style", "text-decoration-thickness",
    "text-indent", "text-transform", "text-underline-offset", "vertical-align",
    "white-space", "widows", "width", "word-break", "word-spacing", "writing-mode",
}
ALLOWED_DISPLAY = {
    "block", "inline", "inline-block", "list-item", "table", "table-row",
    "table-cell", "table-header-group", "table-row-group", "table-footer-group",
    "none", "flex", "inline-flex", "grid", "inline-grid",
}
ALLOWED_POSITION = {"static", "relative"}
SAFE_TAGS = {
    "a", "article", "aside", "blockquote", "body", "caption", "code", "dd", "div",
    "dl", "dt", "em", "figcaption", "figure", "h1", "h2", "h3", "h4", "h5", "h6",
    "hr", "html", "img", "li", "main", "mark", "nav", "ol", "p", "pre", "section",
    "span", "strong", "sub", "sup", "table", "tbody", "td", "th", "thead", "tr",
    "ul", "var",
}
STRUCTURAL_PSEUDOS = {":root", ":first-child", ":last-child", ":only-child", ":empty"}


def audit_stylesheet(input_, limits=None):
    limits = limits or DEFAULT_LIMITS
    violations = []
    audit_breakouts(input_, violations)
    for rule in parse_rules(input_):
        audit_rule(rule, limits, violations)
    return dedupe(violations)


def sanitize_stylesheet(input_, limits=None):
    limits = limits or DEFAULT_LIMITS
    audit = audit_stylesheet(input_, limits)
    if len(audit) == 0:
        return escape_style_text(input_)
    if any(v["message"] == "Forbidden executable or breakout content in NODS." for v in audit):
        return "/* NODX-E027: blocked unsafe style content */"
    out = []
    for rule in parse_rules(input_):
        rule_audit = []
        audit_rule(rule, limits, rule_audit)
        if any(v["severity"] == "error" for v in rule_audit):
            out.append("/* NODX-E027: forbidden NODS rule omitted */")
        else:
            out.append(escape_style_text(rule["source"]))
    return "".join(out) if out else "/* NODX-E027: blocked unsafe style content */"


def yaml_style_to_css(input_):
    rules = []
    selector = ""
    declarations = []

    def flush_rule():
        nonlocal selector, declarations
        if selector and declarations:
            rules.append(selector + " { " + " ".join(declarations) + " }")
        selector = ""
        declarations = []

    for raw in input_.replace("\r\n", "\n").split("\n"):
        if not raw.strip() or raw.lstrip().startswith("#"):
            continue
        if raw.startswith("\t"):
            raise ValueError("YAML style indentation must use spaces.")
        indent = len(raw) - len(raw.lstrip())
        line = raw.rstrip()
        if indent == 0:
            flush_rule()
            if not line.endswith(":"):
                raise ValueError("YAML style top-level entries must end with `:`.")
            selector = line[:-1].strip()
            if not selector:
                raise ValueError("Invalid YAML style selector.")
        elif selector:
            i = line.find(":")
            if i <= 0:
                raise ValueError("YAML style declaration must use `property: value`.")
            prop = line[:i].strip()
            value = line[i + 1 :].strip()
            if not prop or not value:
                raise ValueError("YAML style declarations require a property and value.")
            declarations.append(f"{prop}: {value};")
        else:
            raise ValueError("Invalid YAML style structure.")
    flush_rule()
    return "\n".join(rules)


def parse_rules(input_):
    rules = []
    start = 0
    while start < len(input_):
        open_ = input_.find("{", start)
        if open_ < 0:
            break
        depth = 1
        i = open_ + 1
        while i < len(input_) and depth > 0:
            if input_[i] == "{":
                depth += 1
            elif input_[i] == "}":
                depth -= 1
            i += 1
        if depth != 0:
            rules.append({"prelude": input_[start:open_], "declarations": input_[open_ + 1 :], "source": input_[start:]})
            break
        rules.append({"prelude": input_[start:open_], "declarations": input_[open_ + 1 : i - 1], "source": input_[start:i]})
        start = i
    return rules


def audit_breakouts(input_, violations):
    decoded = decode_css_escapes(input_).lower()
    for construct in ("</style", "<script", "<svg", "<iframe", "<object", "<embed", "vbscript:", "expression(", "@import"):
        if construct in decoded:
            violations.append(err(construct, "Forbidden executable or breakout content in NODS."))


def audit_rule(rule, limits, violations):
    prelude = rule["prelude"].strip()
    if prelude.startswith("@"):
        audit_at_rule(prelude, violations)
        lower = prelude.lower()
        if lower.startswith("@media") or lower.startswith("@supports"):
            for nested in parse_rules(rule["declarations"]):
                audit_rule(nested, limits, violations)
            return
    else:
        audit_selector(prelude, violations)
    audit_declarations(rule["declarations"], limits, violations)


def audit_at_rule(prelude, violations):
    lower = prelude.lower()
    if lower.startswith("@page") or lower.startswith("@media") or lower.startswith("@supports"):
        return
    violations.append(err(prelude, "Forbidden NODS at-rule."))


def audit_selector(selector, violations):
    if selector == "":
        violations.append(err("selector", "Missing NODS selector."))
        return
    for part in selector.split(","):
        trimmed = part.strip()
        if trimmed == "":
            violations.append(err("selector", "Empty NODS selector list entry."))
            continue
        for token in re.split(r"\s+", trimmed):
            audit_selector_token(token, violations)


def audit_selector_token(token, violations):
    if token == "":
        violations.append(err("selector", "Empty NODS selector token."))
        return
    if token in (">", "+", "~") or token in STRUCTURAL_PSEUDOS:
        return
    if token.startswith(":not(") and token.endswith(")"):
        audit_selector_token(token[5:-1], violations)
        return
    if "::" in token:
        violations.append(err(token, "Pseudo-elements are not allowed in NODS."))
        return
    if ":" in token:
        violations.append(err(token, "Interactive NODS pseudo-class is forbidden."))
        return
    if "*" in token or "|" in token:
        violations.append(err(token, "Forbidden NODS selector."))
        return
    bracket_start = token.find("[")
    if bracket_start >= 0:
        bracket_end = token.find("]")
        if bracket_end < 0:
            violations.append(err(token, "Unterminated attribute selector."))
            return
        base = token[:bracket_start]
        attr = token[bracket_start + 1 : bracket_end]
        if base != "" and not is_safe_selector_base(base):
            violations.append(err(base, "Unsupported NODS selector."))
            return
        if not is_safe_attribute_selector(attr):
            violations.append(err(attr, "Forbidden NODS attribute selector."))
        return
    if not is_safe_selector_base(token):
        violations.append(err(token, "Unsupported NODS selector."))


def is_safe_selector_base(token):
    if token == "":
        return False
    idx = first_index_of_any(token, [".", "#"])
    if idx == 0:
        rest = token[1:]
        next_ = first_index_of_any(rest, [".", "#"])
        ident = rest if next_ < 0 else rest[:next_]
        if not is_ident(ident):
            return False
        if next_ < 0:
            return token[0] in (".", "#")
        return is_safe_selector_base(rest[next_:])
    if idx > 0:
        return is_safe_tag(token[:idx]) and is_safe_selector_base(token[idx:])
    return is_safe_tag(token)


def first_index_of_any(token, chars):
    indexes = [token.find(ch) for ch in chars if token.find(ch) >= 0]
    return min(indexes) if indexes else -1


def is_safe_tag(token):
    return token in SAFE_TAGS


def is_safe_attribute_selector(attr):
    i = 0
    name = ""
    while i < len(attr) and re.match(r"[A-Za-z0-9_-]", attr[i]):
        name += attr[i]
        i += 1
    if name == "" or not is_ident(name):
        return False
    rest = attr[i:].strip()
    if rest == "":
        return name in {"lang", "dir", "role", "data-tag", "data-tone", "data-variant", "data-color", "data-status"}
    for op in ("~=", "|=", "^=", "$=", "*=", "="):
        if rest.startswith(op):
            value = rest[len(op) :].strip()
            if ((value.startswith('"') and value.endswith('"')) or (value.startswith("'") and value.endswith("'"))) and len(value) >= 2:
                inner = value[1:-1]
                return len(inner) > 0 and re.search(r"""["'<>]""", inner) is None
            return False
    return False


def audit_declarations(declarations, limits, violations):
    for declaration in declarations.split(";"):
        trimmed = declaration.strip()
        if trimmed == "":
            continue
        idx = trimmed.find(":")
        if idx < 0:
            violations.append(err(trimmed, "Malformed NODS declaration."))
            continue
        prop = trimmed[:idx].strip().lower()
        value = trimmed[idx + 1 :].strip()
        if prop in {"behavior", "-moz-binding", "-ms-behavior", "binding"}:
            violations.append(err(prop, "Forbidden NODS property."))
            continue
        if not (prop.startswith("--") and len(prop) > 2) and prop not in ALLOWED_PROPERTIES:
            violations.append(warn(prop, "Unsupported NODS property."))
            continue
        audit_value(prop, value, limits, violations)


def audit_value(prop, value, limits, violations):
    decoded = decode_css_escapes(value)
    lower = decoded.lower()
    for construct in ("expression(", "javascript:", "vbscript:", "@import", "behavior:"):
        if construct in lower:
            violations.append(err(construct, "Forbidden NODS construct in value."))
    for url in style_urls(value):
        if not classify_uri(ReferenceKind.Style, url, limits)["ok"]:
            violations.append(err(url, "Unsafe NODS URL."))
    if prop == "position" and lower not in ALLOWED_POSITION:
        violations.append(err(value, "Forbidden NODS positioning value."))
    if prop == "display" and lower not in ALLOWED_DISPLAY:
        violations.append(err(value, "Forbidden NODS display value."))


def style_urls(input_):
    urls = []
    lower = input_.lower()
    offset = 0
    while True:
        start = lower.find("url(", offset)
        if start < 0:
            break
        url_start = start + 4
        end = input_.find(")", url_start)
        if end < 0:
            break
        inner = input_[url_start:end].strip()
        if (inner.startswith('"') and inner.endswith('"')) or (inner.startswith("'") and inner.endswith("'")):
            inner = inner[1:-1]
        urls.append(inner)
        offset = end + 1
    return urls


def decode_css_escapes(input_):
    out = ""
    i = 0
    while i < len(input_):
        if input_[i] == "\\" and i + 1 < len(input_):
            j = i + 1
            hex_ = ""
            while j < len(input_) and len(hex_) < 6 and re.match(r"[0-9a-fA-F]", input_[j]):
                hex_ += input_[j]
                j += 1
            if hex_:
                if j < len(input_) and input_[j] in " \t\n":
                    j += 1
                out += chr(int(hex_, 16))
                i = j
                continue
            if j < len(input_):
                out += input_[j]
                i = j + 1
                continue
        out += input_[i]
        i += 1
    return out


def escape_style_text(input_):
    return input_.replace("<", "\\3C ").replace(">", "\\3E ")


def dedupe(list_):
    seen = set()
    result = []
    for v in list_:
        key = v["construct"] + "\0" + v["message"] + "\0" + v["severity"]
        if key not in seen:
            seen.add(key)
            result.append(v)
    return result


def err(construct, message):
    return {"construct": construct.strip(), "message": message, "severity": "error"}


def warn(construct, message):
    return {"construct": construct.strip(), "message": message, "severity": "warning"}


def is_ident(name):
    return re.match(r"^[A-Za-z_][A-Za-z0-9_-]*$", name) is not None


auditStylesheet = audit_stylesheet
sanitizeStylesheet = sanitize_stylesheet
yamlStyleToCss = yaml_style_to_css
